proxy/credentials_issuer_headers.go

```diff
@@ -0,0 +1,72 @@
+package proxy
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"text/template"
+
+	"github.com/ory/oathkeeper/rule"
+	"github.com/pkg/errors"
+)
+
+type CredentialsHeadersConfig map[string]string
+
+type CredentialsHeaders struct {
+	RulesCache *template.Template
+}
+
+func NewCredentialsIssuerHeaders() *CredentialsHeaders {
+	return &CredentialsHeaders{
+		RulesCache: template.New("rules").
+			Option("missingkey=zero").
+			Funcs(template.FuncMap{
+				"print": func(i interface{}) string {
+					if i == nil {
+						return ""
+					}
+					return fmt.Sprintf("%v", i)
+				},
+			}),
+	}
+}
+
+func (a *CredentialsHeaders) GetID() string {
+	return "headers"
+}
+
+func (a *CredentialsHeaders) Issue(r *http.Request, session *AuthenticationSession, config json.RawMessage, rl *rule.Rule) error {
+	if len(config) == 0 {
+		config = []byte("{}")
+	}
+
+	var cfg CredentialsHeadersConfig
+	d := json.NewDecoder(bytes.NewBuffer(config))
```

Comment:
The `CredentialsHeadersConfig` is just a `map[string]string` so there are no unknown fields, the map keys correspond to HTTP headers so they have to remain freeform. Thinking about it now, though, I realize that may constrain future configuration options. I could turn it into a struct with a `Headers map[string]string` field instead, which would go from

```json
"credentials_issuer": {
    "handler": "headers",
    "config": {
        "X-User": "{{ print .Subject }}",
        "X-Audience": "{{ print .Extra.aud }}"
    }
}
```
to
```json
"credentials_issuer": {
    "handler": "headers",
    "config": {
        "headers": {
            "X-User": "{{ print .Subject }}",
            "X-Audience": "{{ print .Extra.aud }}"
        }
    }
}
```

But that may end up being an unnecessary change if we never introduce any new configuration options for that issuer. I can't think of anything I could add, though. I'll let you decide which you prefer and implement as directed.

Reply:
Hm, it could be possible that we want to configure the headers handler further (maybe add a `filter:true/false` option which removes unused X- headers?). I think it's a good idea to separate the template from potential future updates to the config. It will improve BC as well!

Reply:
Done :ok_hand: